File: src/main/java/co/uk/jiveelection/campaign/output/twitter/TwitterOutput.java
package co.uk.jiveelection.campaign.output.twitter;

import co.uk.jiveelection.campaign.jive.Jive;
import co.uk.jiveelection.campaign.output.Output;
import com.google.auto.factory.AutoFactory;
import com.google.auto.factory.Provided;
import twitter4j.Status;
import twitter4j.Twitter;
import twitter4j.TwitterException;

/**
 * Configures a Twitter user to tweet on behalf of.
 */
@AutoFactory
public class TwitterOutput implements Output<Status> {

    private static final int TWEET_CHARACTER_LIMIT = 280;
    private static final int TWEET_PAGINATION_INDEX = 265;

    private final Jive jiveBot;
    private final Twitter twitterOutput;

    /**
     * Constructs a new TwitterOutput.
     *
     * @param jiveBot Jive: The Jive bot to attach to
     * @param twitter Twitter: Used to tweet statuses
     */
    public TwitterOutput(Jive jiveBot, @Provided Twitter twitter) {
        this.jiveBot = jiveBot;
        this.twitterOutput = twitter;
    }

    @Override
    public void outputJive(String jive) {
        // Tweet, xzibit style
        // Check if translator is > tweet character limit
        // if yes break into smaller tweet with [1/2], [2,2] suffix
        // if not tweet
        if (jive.length() > TWEET_CHARACTER_LIMIT) {
            int i = jive.lastIndexOf(" ", TWEET_PAGINATION_INDEX);

            if (i != -1) {
                String first = jive.substring(0, i) + " [1/2]";
                String second = jive.substring(i + 1) + " [2/2]";

                onOutputComplete(tweet(first));
                onOutputComplete(tweet(second));
            }
        } else {
            onOutputComplete(tweet(jive));
        }
    }

    @Override
    public void onOutputComplete(Status status) {
        jiveBot.onJiveComplete(status);
    }

    private Status tweet(String jive) {
        Status status = null;

        try {
            status = twitterOutput.updateStatus(jive);
        } catch (TwitterException e) {
            e.printStackTrace();
            System.out.println(jive);
        }

        return status;
    }
}

File: src/main/kotlin/co/uk/jiveelection/campaign/output/twitter/TwitterOutput.kt
package co.uk.jiveelection.campaign.output.twitter

import co.uk.jiveelection.campaign.jive.Jive
import co.uk.jiveelection.campaign.output.Output
import com.google.auto.factory.AutoFactory
import com.google.auto.factory.Provided
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import twitter4j.Status
import twitter4j.Twitter

/**
 * Configures a Twitter user to tweet on behalf of.
 */
@AutoFactory
class TwitterOutput(private val jiveBot: Jive<*, *, *>, @Provided private val twitter: Twitter) : Output<Status> {

    private val logger: Logger = LoggerFactory.getLogger(TwitterOutput::class.java)

    override fun outputJive(jive: String) {
        // Tweet, xzibit style
        // Check if translator is > tweet character limit
        // if yes break into smaller tweet with [1/2], [2,2] suffix
        // if not tweet
        if (jive.length > TWEET_CHARACTER_LIMIT) {
            val i = jive.lastIndexOf(" ", TWEET_PAGINATION_INDEX)

            if (i != -1) {
                val first = jive.substring(0, i) + " [1/2]"
                val second = jive.substring(i + 1) + " [2/2]"

                onOutputComplete(tweet(first))
                onOutputComplete(tweet(second))
            }
        } else {
            onOutputComplete(tweet(jive))
        }
    }

    override fun onOutputComplete(status: Status) {
        jiveBot.onJiveComplete(status)
        logger.info("Sent jive status from: ${twitter.screenName}")
    }

    private fun tweet(jive: String): Status {
        return twitter.updateStatus(jive)
    }

    companion object {
        private const val TWEET_CHARACTER_LIMIT = 280
        private const val TWEET_PAGINATION_INDEX = 265
    }
}
